simde/x86/avx2.h

```diff
@@ -4714,7 +4714,7 @@ simde_mm256_srl_epi16 (simde__m256i a, simde__m128i count) {
       uint64_t shift = HEDLEY_STATIC_CAST(uint64_t , (count_.i64[0] > 16 ? 16 : count_.i64[0]));
 
       #if defined(SIMDE_VECTOR_SUBSCRIPT_SCALAR)
-        r_.u16 = a_.u16 >> HEDLEY_STATIC_CAST(int16_t, shift);
+        r_.u16 = a_.u16 >> HEDLEY_STATIC_CAST(uint16_t, shift);
```

Comment:
These are potentially bad for other architectures, since usually the shift count is signed (for example, [NEON](https://developer.arm.com/architectures/instruction-sets/simd-isas/neon/intrinsics?search=vshrq_n_u16) or [SSE](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_srli_epi16&expand=5507)).  OTOH, it's good for PPC (`vec_srl` uses unsigned).

So this should probably be something like

```c
#if defined(SIMDE_ARCH_E2K) || defined(SIMDE_ARCH_POWER)
  r_.u16 = a_.u16 >> HEDLEY_STATIC_CAST(uint16_t, shift);
#else
  r_.u16 = a_.u16 >> HEDLEY_STATIC_CAST(int16_t, shift);
#endif
```

Or, even better, we could do something like this:

```c
#if defined(SIMDE_ARCH_E2K) || defined(SIMDE_ARCH_POWER)
  #define SIMDE_VECTOR_SHIFT_PREFER_UNSIGNED
#endif
```

In simde-common.h instead of copying the logic everywhere.

Reply:
I see. But won't this `#if/#else/#endif` block be too big for each place where supposed change is to be introduced? What about something like this:
```
#if defined(SIMDE_ARCH_E2K) || defined(SIMDE_ARCH_POWER)
  #define SIMDE_VECTOR_SHIFT_TYPE_8 uint8_t
  #define SIMDE_VECTOR_SHIFT_TYPE_16 uint16_t
  #define SIMDE_VECTOR_SHIFT_TYPE_32 uint32_t
#else
  #define SIMDE_VECTOR_SHIFT_TYPE_8 int8_t
  #define SIMDE_VECTOR_SHIFT_TYPE_16 int16_t
  #define SIMDE_VECTOR_SHIFT_TYPE_32 int32_t
#endif
```
and then in cases like as one discussed,  simde/x86/avx2.h:4717, we do the following:
```
   r_.u32 = a_.u32 >> HEDLEY_STATIC_CAST(SIMDE_VECTOR_SHIFT_TYPE_16, shift);
```
and in cases like as in simde/x86/sse2.h:5879, this:
```
 r_.u16 = a_.u16 >> HEDLEY_STATIC_CAST(SIMDE_VECTOR_SHIFT_TYPE_8, imm8);
```
What do you think about it?

Reply:
I like that, though I'd be inclined to combine the cast into the macro; something like

```
#if defined(SIMDE_ARCH_E2K) || defined(SIMDE_ARCH_POWER)
  #define SIMDE_CAST_VECTOR_SHIFT_COUNT(width, value) HEDLEY_STATIC_CAST(uint##width##_t, (value))
#else
  #define SIMDE_CAST_VECTOR_SHIFT_COUNT(width, value) HEDLEY_STATIC_CAST(int##width##_t, (value))
#endif
```

Reply:
That seems to be the best solution. Implemented in 11f5827.